CLI Compatibility for Linux

jfgordon2 opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #32 

## Description

- Added github shim script to launch electron app from CLI
- Added check for linux environment to existing CLI file
- Adds static file directory to $PATH after install (to capture the github shim)
- BUGFIX: added check for linux-only editors
- Added Elementary Code as an editor as well as Elementary's Terminal
Tested in Elementary OS 5.1.2

### Screenshots

![image](https://user-images.githubusercontent.com/55799997/77005456-d0294880-692e-11ea-8615-039756a12c95.png)

## Release notes

- Added support for CLI for linux
- Added support for Elementary OS Terminal/Code

Notes:

jfgordon2 committed:
Added linux CLI

Added github.sh script to launch electron app from CLI, and added linux environment to CLI file; tested successfully in Elementary OS 5.1.2; still needs modification on linux-after-install script for PATH

jfgordon2 committed:
Updated Linux github CLI

Removed restriction on electron CLI for linux, added shim to find binary, and added to path; also bugfixed check for editors to also include linux enums to allow linux-only editors to not error out, and added Elementary Code and Elementary Terminal

jfgordon2 committed:
Clean-Up

Removed value set during testing to prevent conflict on install

shiftkey commented on app/src/cli/open-desktop.ts:
```diff
@@ -12,10 +12,12 @@ export function openDesktop(url: string = '') {
     throw new Error(
       `Desktop command line interface not currently supported on platform ${
-        process.platform
+      process.platform
```
This change is introducing a linter warning, because Prettier Has Opinions:tm: 

```suggestion
        process.platform
```

shiftkey commented on script/linux-after-install.sh:
```diff
@@ -4,8 +4,10 @@ set -e
 
 PROFILE_D_FILE="/etc/profile.d/github-desktop.sh"
 INSTALL_DIR="/opt/${productFilename}"
+CLI_DIR="$INSTALL_DIR/resources/app/static"
```
Because we have full control over how the app is built for packaging, I think we can make the `github` shortcut live wherever we like. Other OSes have their own constraints:

 - Windows includes the versions in the path, so we need to update the shortcut after launch
 - macOS may requires elevation to install it at a predictable location, so we have an in-app flow to check and prompt

I'm going to leave this here for now because it's working, but if I have any better ideas I'll let you know.

jfgordon2 replied:
AH that would certainly make it easier; then we just can create a symbolic link in the /usr/bin directory
ln -s $INSTALL_DIR/resources/app/static/github /usr/bin/github

shiftkey commented on app/static/linux/github:
```diff
@@ -0,0 +1,28 @@
+if [ -f "$GITHUB_PATH/github-desktop-dev" ]; then
+	BINARY_NAME="github-desktop-dev"
+else
+  BINARY_NAME="github-desktop"
```
Change this to a tab to align it with the contents of `if`:

```suggestion
	BINARY_NAME="github-desktop"
```

shiftkey reviewed (commented):
Thanks for starting this @jfgordon2!

I wanted to sit down and test this myself to confirm the CLI launches as expected, but otherwise this is looking good!

jfgordon2 committed:
Update app/src/cli/open-desktop.ts

Co-Authored-By: Brendan Forster <brendan@github.com>

jfgordon2 committed:
Update app/static/linux/github

Co-Authored-By: Brendan Forster <brendan@github.com>

shiftkey commented on app/src/lib/editors/linux.ts:
```diff
@@ -11,6 +11,7 @@ export enum ExternalEditor {
   SublimeText = 'Sublime Text',
   Typora = 'Typora',
   SlickEdit = 'SlickEdit',
+  ElementaryCode = 'Code',
```
Should this be a more explicit label?

```suggestion
  ElementaryCode = 'Elementary Code',
```

jfgordon2 replied:
I had initially considered that since I think "Code" is more universally used in reference to VS Code, but in the install files, in their the AppCenter, in the menu, and on their github, it's just called  "Code."  Since this is just the user-displayed text in the app, it seemed like it'd be easily recognized to the user that does have it installed by the name "Code," and also seem more of a first-class integration by not inadvertently re-naming their application.

shiftkey replied:
👍 these names will also only be visible if the app is installed, so we can assume some familiarity with it if they're on ElementaryOS

shiftkey commented on app/static/linux/github:
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
```
Currently this file does not have the executable bit set after install:

```
$ ls -l "/opt/GitHub Desktop/resources/app/static"
total 960
...
-rw-r--r-- 1 root root    760 Mar 20 11:14 github
...
```

Should we just `chmod a+x` this file so everyone can launch it?

jfgordon2 replied:
ah, good catch! I might have done that without thinking or assumed it was done on the folder level on installation; yes, we'd want to make it executable

shiftkey reviewed (commented):
This is feeling really close @jfgordon2 - I've been able to test it out and I can confirm it works on my end on Ubuntu 19.10. Once we've addressed the last couple of things I'll pull this into an update after I've pushed 2.4.0 out to everyone.

shiftkey commented:
@jfgordon2 don't worry about the warning here about conflicts - once you've applied that fix to the file mode for the script I'm happy to figure out how to rewrite the history so the diff is easier to follow.

jfgordon2 committed:
Added binary to /usr/bin

Added binary to /usr/bin to avoid having to add to PATH

jfgordon2 commented:
@shiftkey I probably should have just rebased this to remove the confusing commits; I've updated this to add file permissions and removed your `profile.d` script which put the binaries on the PATH variable, and replaced it with symlinks directly to the files in the `/usr/bin` directory

shiftkey commented:
@jfgordon2 I've rebased to drop those unnecessary commits, should be good to land once this goes green. Then I'll cherry-pick it over to the release branch so it can go out with #237.

shiftkey reviewed (approved):
Thanks for working with me to get this supported @jfgordon2. 

Can't wait to get this out for others to enjoy!
